qemu/tests/virtio_serial_empty_line.py

```diff
@@ -0,0 +1,114 @@
+    return device_name.strip("../\n")
+
+
+def check_port_info(session, device_name, check_options):
+    """
+    Check the port info per given options
+    :param session: guest session
+    :param device_name: the device name of the port
+    :param check_options: the options and values to be checked
+    :return False if no such device in port info path; or dictionary
+     including unmatched options: {option: [real_value, exp_value]}
+    """
+    port_info = get_port_info(session).get(device_name)
+    if not port_info:
+        return False
+    unmatch_items = {}
+    for option, exp_value in check_options.items():
+        real_value = port_info.get(option)
+        if exp_value != real_value:
+            unmatch_items.update({option: [real_value, exp_value]})
+    return unmatch_items
+
+
+def get_port_info(session):
+    """
+    Get all the kernel debug info of virtio port
+    :param session: guest session
+    :return the dictionary of all port infos: {port_dev: {option: value}}
+    """
+    virtio_ports_debug_path = "/sys/kernel/debug/virtio-ports/"
+    info_dict = {}
+    port_devs = session.cmd_output_safe(
+        "ls %s" % virtio_ports_debug_path).split()
+    for port_dev in port_devs:
+        port_infos = session.cmd_output("cat %s%s" % (
+            virtio_ports_debug_path, port_dev)).splitlines()
+        port_dict = {}
+        for line in port_infos:
+            option, value = line.split(':')
+            port_dict.update({option: value.strip()})
+        info_dict.update({port_dev: port_dict})
+    return info_dict
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    KVM virtio_console test
+
+    1) Start guest with virtio-serial device
+    2) Send empty line('\n') from host to guest or vice versa
+    3) Check inside guest by "cat /sys/kernel/debug/virtio-ports/vport*p*"
+       bytes_sent or bytes_received increased by 1
+
+    :param test: kvm test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment
+    """
+    virtio_test = VirtioPortTest(test, env, params)
+
+    (vm, guest_worker, port) = virtio_test.get_vm_with_single_port()
+    session = vm.wait_for_login()
+
+    device_name = get_device_name_of_port(session, port.name)
+    try:
+        if params["sender"] == "guest":
+            # Send empty line('\n') from guest to host
+            port.open()
+            # 'echo' automatically adds '\n' in the end of each writing
+            send_data_command = 'echo "" > /dev/%s' % device_name
+            session.cmd(send_data_command, timeout=120)
+            received_data = port.sock.recv(10)
+            if received_data != b"\n":
+                test.fail("Received data is not same as the data sent,"
+                          " received %s, while expected '\n'"
+                          % received_data)
+            check_option = {"bytes_sent": "1"}
```

Comment:
@yanan-fu I guess no, and I suppose it could be used by other modules in the future, by utilize the function `check_port_info(session, device_name, check_options)`, but won't be here in this script. So, that does not matter where `check_option` is defined.

Reply:
@yanan-fu And another point is, this parameter should be only configurable for the function, not configurable for the user, so prefer defining it here in the script.

Reply:
@vivianQizhu  make sense, thanks !

Could you please provide the latest test result ?  Others LGTM.